colossalai/utils/model/lazy_init_context.py

```diff
@@ -36,126 +38,128 @@ class LazyInitContext():
         self._intercepted_init_func_cache = []
         self._nn_init_methods = self._get_nn_init_methods()
         self._torch_mod_cls = torch.nn.modules.module.Module
-        
+
         if extra_torch_tensor_func:
             # use tuple to remove duplicates
             self._torch_tensor_funcs = tuple(self.tensor_set_value_func + extra_torch_tensor_func)
         else:
             self._torch_tensor_funcs = self.tensor_set_value_func
-    
+
     def _cache_func(self, func):
         """
         This method wraps the ``torch.nn.init`` method so that the function call
         is cached instead of being executed.
         """
+
         def wrapped_init_func(*args, **kwargs):
             self._intercepted_init_func_cache.append(dict(func=func, args=args, kwargs=kwargs))
+
         return wrapped_init_func
-            
+
     def _get_nn_init_methods(self):
         """
         This method looks for all available functions in the ``torch.nn.init``
         module.
         """
         nn_init_method_names = dir(torch.nn.init)
         nn_init_methods = []
-        
+
         # look for all methods in ``torch.nn.init`` module
         for name in nn_init_method_names:
             nn_init_methods.append((name, getattr(torch.nn.init, name)))
-            
+
         def _has_tensor_in_arg(func):
-            hints = typing.get_type_hints(torch.nn.init.normal_)
+            hints = typing.get_type_hints(func)
             for k, v in hints.items():
                 if v is torch.Tensor:
                     return True
             return False
-        
+
         def _is_init_method(item):
             name, func = item
-            if (not isinstance(func, types.FunctionType) or
-                name.startswith('_') or
-                not name.endswith('_') or
-                not _has_tensor_in_arg(func)):
+            if (not isinstance(func, types.FunctionType) or name.startswith('_') or not name.endswith('_')
+                    or not _has_tensor_in_arg(func)):
                 return False
             else:
                 return True
-            
+
         # remove methods which are not init functions
         nn_init_methods = list(filter(_is_init_method, nn_init_methods))
         return nn_init_methods
-    
+
     def _wrap_module_init(self, func):
         """
         This method wraps the calls to the `__init__` of ``torch.nn.Module`` and replaces
         the argument device with value 'meta' so that all modules are created as meta tensors.
         """
         has_device = 'device' in inspect.signature(func).parameters
 
-        def layer_lazy_init(*args, **kwargs):
-            if has_device:
-                kwargs['device'] = 'meta'
-            func(*args, **kwargs)
+        def layer_lazy_init(module, *args, **kwargs):
+            self._intercepted_init_func_cache.append(dict(func=func, module=module, args=args, kwargs=kwargs))
+            func(module, *args, **kwargs)
+            module.to('meta')
```

Comment:
I am wondering whether this will take the same amount of time as just init the model like normal. If so, I think we can check if kwargs contain `device`, if yes, we set `device` to `meta`, if no, instantiate it and convert to `meta`. This should speed up the process.

Reply:
I see. I will make module with device kwargs by pass the normal init.